<div style="border:solid steelblue 3px; padding: 20px; border-radius: 10px">

**Александр, привет!** 👋

Меня зовут Алексей Манелюк, я буду проверять твой проект. В Яндекс Практикуме принято общаться на "ты", но если некомфортно - скажи и перейдем на "Вы".

Моя основная цель - не указать на ошибки, а поделиться опытом и помочь в освоении анализа данных.

В тетрадке ты увидишь мои комментарии, пожалуйста, не удаляй, не перемещай и не изменяй их. У комментариев могут быть приписки **V2**, **V3**, **V...**, которые соответствуют каждой версии проверки.

Я буду использовать цветовую разметку:
<p>
<div class="alert-danger" style="padding: 20px; border-radius: 10px">
<b>Комментарий ревьюера ❌</b>

Так выделены критические замечания. Без их отработки я не смогу принять проект. </div>

<div class="alert-warning" style="padding: 20px; border-radius: 10px">
<b>  Комментарий ревьюера ⚠️</b>

Так выделены небольшие замечания или полезные советы. Постарайся учесть их.</div>
<p>
<div class="alert-success" style="padding: 20px; border-radius: 10px">
<b>Комментарий ревьюера ✔️</b>

Так выделены правильные решения.</div>

Предлагаю работать в диалоге: если что-то дорабатываешь или отвечаешь на мой комментарий - пиши об этом. Мне будет удобней, если ты выделишь свои комментарии:
<p>
<div class="alert-info" style="padding: 20px; border-radius: 10px">
<b>Комментарий студента 💎</b>

Здесь исправил.</div>

Все это поможет мне быстрее принять твой проект.</div>

<div class="alert-info" style="padding: 20px; border-radius: 10px">
<b>Комментарий студента 💎</b>

Ты можешь написать свой комментарий, если скопируешь этот блок в ячейку Markdown.</div>

# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



<div class="alert-success" style="padding: 20px; border-radius: 10px">
<b>Комментарий ревьюера ✔️</b>

Несмотря на то, что первая часть этого проекта проверяется автоматизрованно, дам по ней пару советов - они пригодятся тебе в следующих проектах.

📌 Начать проект стоит с описания (цели исследования, описание данных и краткий план работы), а каждый раздел закончить промежуточным выводом.

Как правильно оформить проект ты можешь прочитать в материалах курса:

`Полезные инструкции для учёбы - Оформление проекта - Рекомендации по выполнению проектов`

## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)
data['purpose_category']

0        операции с недвижимостью
1          операции с автомобилем
2        операции с недвижимостью
3           получение образования
4              проведение свадьбы
                   ...           
21520    операции с недвижимостью
21521      операции с автомобилем
21522    операции с недвижимостью
21523      операции с автомобилем
21524      операции с автомобилем
Name: purpose_category, Length: 21331, dtype: object

### Шаг 3. Исследуйте данные и ответьте на вопросы

**Предварительная оценка заданий**

Вопросы, заданные в пунктах 2.7.1 - 2.7.4 имеют схожую структуру, и для ответа на них требуется расчет одного 
и того же показателя - 
процентного отношения невыплаченных в срок кредитов к общему количеству кредитов. Отличаются только категориии заемщиков.

В таком случае целесообразно создать функцию, в аргументе которой указывается столбец датафрейма, содержащий показатель, 
по которому необходимо категорирование.
**Функция должна:**
- группировать датафрейм по столбцу, указанному в аргументе функции.
- считать количество просроченных и взятых кредитов по столбцу `debt` для каждой из категорий.
- находить отношение количества просроченных кредитов к количеству взятых кредитов.
- сводить всю получившуюся информацию в единую таблицу.

<div class="alert-info" style="padding: 20px; border-radius: 10px">
<b>Комментарий студента 💎</b>

При разработке новой функции взамен старой обратил внимание, что практически все действия по подготовке данных для анализа можно вместить внутри функции. Надеюсь, это не усложнит процесс проверки, постарался подоробно расписать каждое действие внутри функции.
Свои ячейки, которые стали неактуальны, я удалил, чтобы не путаться.
(На будущее: если какие-то мои действия по итогам ревью оказываются неверными/ненужными, мне их удалять или оставлять?).</div>

<div class="alert-success" style="padding: 20px; border-radius: 10px">
<b>Комментарий ревьюера V2 ✔️</b>

Вообще, закомментированый код лучше удалить из финальной версии проекта. Но для себя ты можешь оставить копию тетрадки. 🙂

In [22]:
# создаем функцию pivot_debt, в аргументе которой будет указываться столбец датафрейма, по которому требуется группировка
def pivot_debt(column):
# объявляем переменную table_debt, которой присваевается сводная таблица, где index берется из аргумента функции pivot_debt,
# значения из столбца debt, а производимые действия - это sum и count.
    table_debt = data.pivot_table(index=column,values='debt',aggfunc=['sum','count'])
# Создаем новый столбец для table_debt, в котором будет соотношение количества должников и количество взятых кредитов.
    table_debt['debtors_to_debts_ratio'] = table_debt['sum']['debt']/table_debt['count']['debt']
# для более понятного и красивого отображения заголвков столбцов датафрейма применим метод reset_index.
    table_debt_res = table_debt.reset_index()
# переименуем столбцы так, чтобы было понятно, какая информация в них указана.
    table_debt_res.rename(columns ={'sum':'debtors','count':'debts','debt':''},inplace=True)
# Отсечем нерепрезентативные категории по признаку количества выданных кредитов. Я установил порог отсева равный 500.
# для генеральной совокупности равной 21331 - это нормальный порог отсева или должен быть ниже?
# Также в этой строке кода отсортируем строки по столбцу debtors_to_debts_ratio, чтобы на первом месте были самые высокорисковые группы.
    table_debt_res_sort = table_debt_res.loc[table_debt_res['debts'] >= 500].sort_values(by='debtors_to_debts_ratio', ascending=False)
# Добавим в датафрейм новый столбец, где соотношение должников к общему количеству кредитов будет отображено в процентах.
    table_debt_res_sort['debtors_to_debts_ratio_percent'] = table_dhttps://k8s.jupyterhub.praktikum-services.ru/user/user-0-1591616925/notebooks/a078c6bc-6195-45c7-a4b2-a60eff4bc866.ipynb#3.1-%D0%95%D1%81%D1%82%D1%8C-%D0%BB%D0%B8-%D0%B7%D0%B0%D0%B2%D0%B8%D1%81%D0%B8%D0%BC%D0%BE%D1%81%D1%82%D1%8C-%D0%BC%D0%B5%D0%B6%D0%B4%D1%83-%D0%BA%D0%BE%D0%BB%D0%B8%D1%87%D0%B5%D1%81%D1%82%D0%B2%D0%BE%D0%BC-%D0%B4%D0%B5%D1%82%D0%B5%D0%B9-%D0%B8-%D0%B2%D0%BE%D0%B7%D0%B2%D1%80%D0%B0%D1%82%D0%BE%D0%BC-%D0%BA%D1%80%D0%B5%D0%B4%D0%B8%D1%82%D0%B0-%D0%B2-%D1%81%D1%80%D0%BE%D0%BA?ebt_res['debtors_to_debts_ratio'].apply('{:.2%}'.format)
    return table_debt_res_sort

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [23]:
children_debt = pivot_debt(data['children'])
children_debt

children debtors  debts debtors_to_debts_ratio  \
                                                   
2        2     194   2052               0.094542   
1        1     444   4808               0.092346   
0        0    1063  14091               0.075438   

  debtors_to_debts_ratio_percent  
                                  
2                          9.45%  
1                          9.23%  
0                          7.54%

<div class="alert-danger" style="padding: 20px; border-radius: 10px">
<b>Комментарий ревьюера ❌</b></h5>

<s>Ты посчитал долю должников в каждой категории заемщиков. Но для сравнения нам также нужно общее количество заемщиков в каждой категории.

Для наглядного сравнения нужно вывести сводную таблицу, в которой будут следующие столбцы:
- название категории
- количество заемщиков (кредитов)
- количество должников (этот столбец опциональный)
- доля просрочки по категории

Вывести такую таблицу можно с использованием методов как `groupby()`, так и `pivot_table()`.

Такие сводные таблицы нужно будет вывести для пунктов 3.1 - 3.4.

Когда ты выведешь такую таблицу, обрати внимание на то, что выборки не сбалансированы, поэтому выявлять зависимости стоит сравнивая самые многочисленные категории заемщиков.</div>

<div class="alert-info" style="padding: 20px; border-radius: 10px">
<b>Комментарий студента 💎</b>
Учел все твои советы, а также добавил столбец с долями просрочки, выраженными в процентах. Так как,на мой взгляд, информация в процентах лучше воспринимается.
</div>

<a id='children'></a>
**Вывод:** Мы получили сводную таблицу, в которой представлены следующие данные:
- `children` — количество детей у заемщика.
- `debtors` — количество должников по кредиту.
- `debts` — количество взятых кредитов.
- `debtors_to_debts_ratio` — отношение количества должников к количеству взятых кредитов.
- `debtors_to_debts_ratio_percent` — отношение количества должников к количеству взятых кредитов в процентах.

По итогам отсева нерепрезентативных категорий (менее 500 выданных кредитов), осталось три категории:
- безедетные заёмщики.
- заёмщики с одним ребенком.
- заёмщики с двумя детьми.

Видно, что должников с 1 ребенком **на 1.69% больше**, чем бездетных должников: **7.54%** от количества бездетных заемщиков и **9.23%** должников от количества заемщиков с 1 ребенком. Также наблюдается тренд на дальнейшее увеличение количества должников в зависимости от количества детей. Так у заемщиков с двумя детьми процент должников составляет **9.45%**, что на **0.22% больше**, чем процент должников среди заемщиков с одним ребенком. Выборки для замещиков с 3 и более детьми оказались нерепрезентативными, в связи с чем невозможно проследить дальнейшее развитие тренда.

<div class="alert-success" style="border-radius: 10px; padding: 20px">
<b>Комментарий ревьюера V2 ✔️</b>

Да, если сравнить только самые многочисленные выборки с 0, 1 и 2 детьми, можно сказать, что чем больше у заемщика детей, тем выше риск допустить просрочку.</div>

<div class="alert-warning" style="padding: 20px; border-radius: 10px">
<b>Комментарий ревьюера ⚠️</b>

<s>А почему ты считаешь, что зависимости нет, ведь процент должников в каждой категории заемщиков разный? Давай будем ориентироваться на то, что на большом кредитном портфеле, разница даже в доли процента будет важна для заказчика.
</div>

<div class="alert-info" style="padding: 20px; border-radius: 10px">
<b>Комментарий студента 💎</b>

Добавил вывод с анлизом разницы для категорий.</div>

[К итогам](#final)

In [24]:
data.loc[data['children'] == 5,'children'].count() #Посмотрим, сколько всего в датафрейме представлено людей, у которых 5 детей.
#С помощью логической индексации отберем все строки с людьми, у которых 5 детей и с помощью метода count() считаем их количество.

9

In [25]:
data['children'].count() #Для сравнения посмотрим общее количество строк с информацией о детях.

21331

**Вывод:** Группа людей, у которых 5 детей, слишком незначительна относительно всего объема данных и поэтому нерепрезентативна. Поэтому при сравнении процента невыплаченных вовремя кредитов данную группу можно не учитывать.

<div class="alert-info" style="padding: 20px; border-radius: 10px">
<b>Комментарий студента 💎</b>

Эти три ячейки уже не нужны, я их оставил, чтобы не нарушать структуру твоих комментариев.</div>

<div class="alert-success" style="padding: 20px; border-radius: 10px">
<b>Комментарий ревьюера ✔️</b>

Согласен, малочисленные категории не стоит учитывать.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [26]:
data['family_status'].unique() # определим количество уникальных семейных положений, чтобы иметь представление, 
#с каким количеством категорий предстоит работать.



array(['женат / замужем', 'гражданский брак', 'вдовец / вдова',
       'в разводе', 'Не женат / не замужем'], dtype=object)

**Вывод:** В переменной `family_status` всего 5 уникальных значений. Дополнительная категоризация не требуется. Можно переходить к определению доли невыплаченных в срок кредитов для каждой категории. 

In [27]:
fam_status_debt = pivot_debt(data['family_status']) # применяем функцию pivot_debt, указав для группировки столбец 'family_status'
fam_status_debt # выводим результат на экран

family_status debtors  debts debtors_to_debts_ratio  \
                                                                 
0  Не женат / не замужем     273   2796               0.097639   
3       гражданский брак     385   4134               0.093130   
4        женат / замужем     927  12261               0.075606   
1              в разводе      84   1189               0.070648   
2         вдовец / вдова      63    951               0.066246   

  debtors_to_debts_ratio_percent  
                                  
0                          9.76%  
3                          9.31%  
4                          7.56%  
1                          7.06%  
2                          6.62%

<div class="alert-warning" style="padding: 20px; border-radius: 10px">
<b>Комментарий ревьюера ⚠️</b>

<s>Добавь пожалуйста количество заемщиков в каждой категории в сводную таблицу.</div>

<div class="alert-info" style="padding: 20px; border-radius: 10px">
<b>Комментарий студента 💎</b>

Добавил.</div>

<div class="alert-success" style="border-radius: 10px; padding: 20px">
<b>Комментарий ревьюера V2 ✔️</b>

Отлично, можно удалить лишний столбец из таблицы.

<a id='children'></a>
**Вывод:** 

Мы получили сводную таблицу, в которой представлены следующие данные:
- `children` — количество детей у заемщика.
- `debtors` — количество должников по кредиту.
- `debts` — количество взятых кредитов.
- `debtors_to_debts_ratio` — отношение количества должников к количеству взятых кредитов.
- `debtors_to_debts_ratio_percent` — отношение количества должников к количеству взятых кредитов в процентах.

Для начала обращаем внимание, что все категории достаточно репрезентативны, чтобы попасть в итоговую выборку.
Также можно заметить, что результаты по категориям разделяются на две группы:
- Люди, которые состоят/состояли в официальном браке.
- Люди, которые не состоят/состояли в официальном браке.

Рассмотрим долю должников для указанных групп.

<div class="alert-warning" style="padding: 20px;
border-radius: 10px">
<b>Комментарий ревьюера ⚠️</b>
    
<s>Как писал выше, разница в 1-2% будет значима для заказчика.

Какую можно выявить зависимость, если обобщить категории по отношению к официальному браку?</div>

<div class="alert-info" style="padding: 20px; border-radius: 10px">
<b>Комментарий студента 💎</b>

Учел твое предложение. Далее идут ячейки с группировкой и расчетами по этому признаку.</div>

In [28]:
def categorize_family_status(row): # создаем функцию, которая категоризирует значения столбца family_status по двум группам.
        if ('не замужем' in row) or ('гражданский брак' in row):
            return 'Холостые люди.'
        else:
            return 'Люди, которые состоят/состояли в официальном браке.'

In [29]:
# Применим функцию categorize_family_status к столбцу family_status датафрейма fam_status_debt.
fam_status_debt['fam_status_category'] = fam_status_debt['family_status'].apply(categorize_family_status)
fam_status_debt

family_status debtors  debts debtors_to_debts_ratio  \
                                                                 
0  Не женат / не замужем     273   2796               0.097639   
3       гражданский брак     385   4134               0.093130   
4        женат / замужем     927  12261               0.075606   
1              в разводе      84   1189               0.070648   
2         вдовец / вдова      63    951               0.066246   

  debtors_to_debts_ratio_percent  \
                                   
0                          9.76%   
3                          9.31%   
4                          7.56%   
1                          7.06%   
2                          6.62%   

                                 fam_status_category  
                                                      
0                                     Холостые люди.  
3                                     Холостые люди.  
4  Люди, которые состоят/состояли в официальном б...  
1  Люди, которые состоят/состояли в официальном б...  
2  Люди, которые состоят/состояли в официальном б...

Мы получили сводную таблицу, в которой представлены следующие данные:
- `family_status` — семейное положение.
- `debtors` — количество должников по кредиту.
- `debts` — количество взятых кредитов.
- `debtors_to_debts_ratio` — отношение количества должников к количеству взятых кредитов.
- `debtors_to_debts_ratio_percent` — отношение количества должников к количеству взятых кредитов в процентах.
- `fam_status_category` — категории людей относительно официального брака.

Посчитаем процент должников для каждой из категории столбца `fam_status_category`.

In [30]:
fam_status_debt.groupby('fam_status_category')['debtors_to_debts_ratio'].mean().apply('{:.2%}'.format)
# находим средний процент должников для каждой из групп

fam_status_category
Люди, которые состоят/состояли в официальном браке.    7.08%
Холостые люди.                                         9.54%
Name: debtors_to_debts_ratio, dtype: object

<a id='family'></a>
**Вывод:** Исходя из полученных данных люди, которые состоят/состояли в официальном браке **на 2.46% "надежнее"** холостых людей (7.08% должников у людей с опытом офицального брака и 9.54% у холостых людей).

<div class="alert-success" style="border-radius: 10px; padding: 20px">
<b>Комментарий ревьюера V2 ✔️</b>

Согласен, заемщики состоящие или состоявшие в браке выглядят более надежными.</div>

[К итогам](#final)

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Уровень дохода — количественная величина, требующая категорирования. Однако в предыдущих заданиях проекта категорирование значений переменной `total_income` уже было осуществлено. Поэтому можно просто воспользоваться им.

In [31]:
data['total_income_category'].unique()# выведем на экран все уникальные значения столбца total_income_category.

array(['B', 'C', 'D', 'E', 'A'], dtype=object)

In [32]:
fam_income_debt = pivot_debt(data['total_income_category'])
fam_income_debt

total_income_category debtors  debts debtors_to_debts_ratio  \
                                                                
2                     C    1353  15921               0.084982   
1                     B     354   5014               0.070602   

  debtors_to_debts_ratio_percent  
                                  
2                          8.50%  
1                          7.06%

<a id='income'></a>
**Вывод:** 
Мы получили сводную таблицу, в которой представлены следующие данные:
- `total_income_category` — категория ежемесячного дохода заемщиков.
- `debtors` — количество должников по кредиту.
- `debts` — количество взятых кредитов.
- `debtors_to_debts_ratio` — отношение количества должников к количеству взятых кредитов.
- `debtors_to_debts_ratio_percent` — отношение количества должников к количеству взятых кредитов в процентах.

При расчетах отсеялись категории `А`, `D`, `E`. Это значит, что количество выданных кредитов для каждой из этих категорий не превышает 500. Указанные категории не будут рассматриваться в дальнейшем ввиду их нерепрезентативности.

Для оставшихся категорий `B ` и `C` получились следующие результаты:
- B (люди с ежемесячным доходом от 200 000 руб. и до 1000 000 руб.) — 7.06% должников.
- C (люди с ежемесячным доходом от 50 000 руб. и до 200 000 руб.)— 8.50% должников.

Это говорит нам о том, что должников с доходом от 50 000 руб.до 200 000 руб. на **1.44% больше** чем должников с зарплатой от 200 000 руб. и до 1000 000 руб. 

Следовательно, можно сделать вывод, что доход заемщика влияет на возврат кредита. Однако для построения тренда нужно больше данных для других категорий заработных плат.

<div class="alert-success" style="border-radius: 10px; padding: 20px">
<b>Комментарий ревьюера V2 ✔️</b>

Да, если сравнить самые большие категории B и C, можно сказать, что заемщики с более высоким доходом платят кредиты вовремя лучше.</div>

<div class="alert-danger" style="padding: 20px; border-radius: 10px">
<b>Комментарий ревьюера ❌</b></h5>

<s>Когда ты выведешь сводную таблицу с количеством заемщиков в каждой категории, ты увидишь, что выборки крайне несбалансированны. Поэтому, для выявления зависимости имеет смысл сравнивать только самые многочисленные выборки.</div>

<div class="alert-info" style="padding: 20px; border-radius: 10px">
<b>Комментарий студента 💎</b>

Действительно, осталось только две категории (если я конечно не установил слишком высокий порог отсева).</div>

[К итогам](#final)

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

Переменная `purpose` является категориальной, однако она заполняется людьми в свободной форме, что вызывает большое количество возможных вариантов значений. В ходе выполнения проекта нами уже проведена категоризация значений указанного столбца. Для поиска ответа на поставленный вопрос можно воспользоваться данной категоризацией.

In [33]:
fam_purpose_debt = pivot_debt(data['purpose_category']) # применяем функцию pivot_debt, указав для группировки столбец 'purpose_category'
fam_purpose_debt # выводим результат на экран

purpose_category debtors  debts debtors_to_debts_ratio  \
                                                                    
0    операции с автомобилем     400   4279               0.093480   
2     получение образования     369   3988               0.092528   
3        проведение свадьбы     183   2313               0.079118   
1  операции с недвижимостью     780  10751               0.072551   

  debtors_to_debts_ratio_percent  
                                  
0                          9.35%  
2                          9.25%  
3                          7.91%  
1                          7.26%

<div class="alert-warning" style="padding: 20px; border-radius: 10px">
<b>Комментарий ревьюера ⚠️</b>

<s>Здесь тоже нужно добавить в таблицу количество заемщиков в каждой категории.</div>

<a id='purpose'></a>
**Вывод:** 
Мы получили сводную таблицу, в которой представлены следующие данные:
- `purpose_category` — категория цели кредита.
- `debtors` — количество должников по кредиту.
- `debts` — количество взятых кредитов.
- `debtors_to_debts_ratio` — отношение количества должников к количеству взятых кредитов.
- `debtors_to_debts_ratio_percent` — отношение количества должников к количеству взятых кредитов в процентах.

Из полученных данных можно сделать вывод, что наиболее рисковоми являются кредиты с целью операций с автомобилем **(9.35% должников)** и получения образования **(9.25% должников)** должников по каждой из категорий. Менее рискованными являются кредиты с целью операций с недвижимостью **(7.26% должников)** и проведения свадебных мероприятий **(7.91% должников)**.

<div class="alert-success" style="border-radius: 10px; padding: 20px">
<b>Комментарий ревьюера V2 ✔️</b>

Всё верно.

<div class="alert-danger" style="padding: 20px; border-radius: 10px">

<b>Комментарий ревьюера ❌</b></h5>

<s>Стоит подсветить для заказчика более надежные и рисковые категории.

<div class="alert-info" style="padding: 20px; border-radius: 10px">
<b>Комментарий студента 💎</b>

Сделано)</div>

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* В исходных данных пропущенные данные есть в двух столбцах: days_employed (общий трудовой стаж в днях) и total_income (ежемесячный доход). Однако за кредитом могут обращаться люди, которые ещё не начали свою трудовую деятельность, и у них пока нет трудового стажа, либо люди, которые в настоящее время не работают и не имеют ежемесячного дохода. Такие люди могут заполнять соответствующие графы в заявке на кредит различными способами: писать "нет", "не работаю", ставить прочерк и т.д., что в свою очередь может вызвать затруднения при переносе информации в электронный вид.

<div class="alert-success" style="padding: 20px; border-radius: 10px">
<b>Комментарий ревьюера ✔️</b>

Предположения о возникновении пропусков правильные: либо человеческий фактор, либо техническая ошибка. </div>

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Заполнение пропусков медианным значением влечет за собой наименьшее искажение данных столбца в целом.

<div class="alert-success" style="padding: 20px; border-radius: 10px">
<b>Комментарий ревьюера ✔️</b>

Всё верно, медиана менее чувствительна к выбросам, чем среднее значение. Для примера можешь рассчитать и сравнить медиану и среднее значение для столбца `days_employed`.</div>

In [34]:
display(data['days_employed'].mean())
display(data['days_employed'].median())

67148.97213258488

1996.732855437927

<div class="alert-info" style="padding: 20px; border-radius: 10px">
<b>Комментарий студента 💎</b>

Разница действительно огромная.</div>

<div class="alert-success" style="border-radius: 10px; padding: 20px">
<b>Комментарий ревьюера V2 ✔️</b>

👍 Да, видно как среднее смещается в сторону выбросов.

### Шаг 4: общий вывод.

<a id='final'></a>
Для ответа на поставленные вопросы мы работали со следующими данными:
- `children` — количество детей в семье.
- `days_employed` — общий трудовой стаж в днях.
- `dob_years` — возраст клиента в годах.
- `education` — уровень образования клиента.
- `education_id` — идентификатор уровня образования.
- `family_status` — семейное положение.
- `family_status_id` — идентификатор семейного положения.
- `gender` — пол клиента.
- `income_type` — тип занятости.
- `debt` — имел ли задолженность по возврату кредитов.
- `total_income` — ежемесячный доход.
- `purpose` — цель получения кредита.

На этапе предобработки нами были обработаны пропуски в столбцах `days_employed` и `total_income`, удалены аномальные значения из столбца `children`, а также удалены дубликаты по всему датафрейму. Также проведена категоризациия данных в столбцах `total_income` и `purpose`.

При рассмотрении категорий заемщиков не учитывались категории, по которым выдано менее 500 кредитов.

По вопросу **"Есть ли зависимость между количеством детей и возвратом кредита в срок?"** мы получили [следующие результаты](#children):

Должников с 1 ребенком **на 1.69% больше**, чем бездетных должников: **7.54%** от количества бездетных заемщиков и **9.23%** должников от количества заемщиков с 1 ребенком. Также наблюдается тренд на дальнейшее увеличение количества должников в зависимости от количества детей. Так у заемщиков с двумя детьми процент должников составляет **9.45%**, что на **0.22% больше**, чем процент должников среди заемщиков с одним ребенком. Выборки для замещиков с 3 и более детьми оказались нерепрезентативными, в связи с чем невозможно проследить дальнейшее развитие тренда.

***Наблюдается зависимость между количеством детей и возвратом кредита в срок. Наличие детей увеличивает риск просрочек по кредиту.***

По вопросу **"Есть ли зависимость между семейным положением и возвратом кредита в срок?"** мы получили [следующие результаты](#family):

Категории семейного положения были разделены на 2 группы:
- Люди,которые состоят/состояли в браке.
- Холостые люди.

По указанным группам можно сделать следующий вывод:

люди, которые состоят/состояли в официальном браке **на 2.46% "надежнее"** холостых людей (7.08% должников у людей с опытом офицального брака и 9.54% у холостых людей).

По вопросу **"Есть ли зависимость между уровнем заработной платы и возвратом кредита в срок?"** мы получили [следующие результаты](#income):

При расчетах отсеялись категории `А` (зарпалата более 1000 000 руб.), `D` (зарплата от 30 000 руб. до 50 000 руб.), `E` (зарпалата до 30 000 руб.). Это значит, что количество выданных кредитов для каждой из этих категорий не превышает 500. Указанные категории не будут рассматриваться в дальнейшем ввиду их нерепрезентативности.

Для оставшихся категорий `B ` и `C` получились следующие результаты:
- B (люди с ежемесячным доходом от 200 000 руб. и до 1000 000 руб.) — 7.06% должников.
- C (люди с ежемесячным доходом от 50 000 руб. и до 200 000 руб.)— 8.50% должников.

Это говорит нам о том, что должников с доходом от 50 000 руб.до 200 000 руб. на **1.44% больше** чем должников с зарплатой от 200 000 руб. и до 1000 000 руб. 

***Следовательно, можно сделать вывод, что доход заемщика влияет на возврат кредита. Чем выше зарплата, тем надежнее заемщик. Однако для построения тренда нужно больше данных для других категорий заработных плат.***

По вопросу **"Есть ли зависимость между целью получения кредита и возвратом кредита в срок?"** мы получили [следующие результаты](#purpose):

Так как цели кредита излагаются заявителями в свободной форме потребовалась дополнительная категоризация данных столбца `purpose`, сформировано 4 категории:

- операции с недвижимостью.
- операции с автомобилем.
- получение образования.
- проведение свадьбы.

Из полученных данных можно сделать вывод, что наиболее рисковоми являются кредиты с целью операций с автомобилем **(9.35% должников)** и получения образования **(9.25% должников)** должников по каждой из категорий. Менее рискованными являются кредиты с целью операций с недвижимостью **(7.26% должников)** и проведения свадебных мероприятий **(7.91% должников)**.

***Наблюдается взаимосвязь между целью кредита и вовзратом кредита в срок.***

По итогам исследования удалось выявить как наиболее рисковые группы *(например холостяк с зарпалтой 45 000 руб., который берет кредит для покупки автомобиля)*, так и наиболее безопасные *(например женатый заемщик без детей и с доходом 150 000 руб., который берет кредит для покупки недвижимости)*

При работе с данными обнаружены достаточно многочисленные пропуски в информации о количестве рабочих дней и заработной плате. Возможно, стоит обратить внимание на процесс сбора и обработки указанных данных.

Также целесообразно предлагать заемщику выбирать цель кредита из предложенных вариантов, оставив возможность свободного ответа только в случае отсутствия подходящего варианта.


<div class="alert-success" style="border-radius: 10px; padding: 20px">
<b>Комментарий ревьюера V2 ✔️</b>

Теперь у тебя получился подробный и правильный вывод, отражающий все основные результаты исследования. 👌</div>

<div class="alert-danger" style="padding: 20px; border-radius: 10px">

<b>Комментарий ревьюера ❌</b></h5>

<s>Общий вывод - важный элемент работы. Его нужно сформулировать, взяв основные результаты из всех шагов проекта. Начать нужно с того, с какими данными мы работали и какие действия провели на этапе предобработки (коротко), далее - сформулировать все выявленные зависимости просрочки от различных факторов, можно дать портрет идеального и рискового заемщика и закончить рекомендациями заказчику по улучшению сбора данных и системы скоринга.

Заказчик, зачастую, читает только общий вывод исследования - поэтому он должен быть максимально информативным.</div>

<div class="alert-info" style="padding: 20px; border-radius: 10px">
<b>Комментарий студента 💎</b>

Готово. Постарался дать максимально развернутый ответ.</div>

<div style="border:solid steelblue 3px; padding: 20px; border-radius: 10px">

<h5><b>🏁 Общий комментарий ревьюера</b></h5>

Ты хорошо поработал над проектом. Но нужно кое-что поправить, чтобы я его принял.

Для удобства, критические замечания я добавил в меню **Contents** слева.

Также, постарайся отработать желтые комментарии.

Буду ждать твой доработанный проект. ⌛️

<div style="border:solid steelblue 3px; padding: 20px; border-radius: 10px">

<h5><b>🏁 Общий комментарий ревьюера V2</b></h5>

Отличная работа! Ты внимательно учёл мои комментарии и проделал большую работу по улучшению проекта.

Успехов в дальнейшем обучении! 🏆

</div>